### ISSUE TITLE
Question on AutoSlim on yoloxnano

### ISSUE THREAD

**Issue by TianpengBu** [open-mmlab/mmrazor#309]
### Checklist

- I have searched related issues but cannot get the expected help.
- I have read related documents and don't know what to do.

### Describe the question you meet

I tried to use autoSlim to prune yoloxnano. The training process can be ran successfully. However, I found the mAP does not converge. 

### My config below:

```
_base_ = [
    '../../../_base_/schedules/mmdet/schedule_1x.py',
    '../../../_base_/mmdet_runtime.py'
]
img_scale = (640, 640)
test_img_scale = (416, 416)
NC = 1 
CLASSES = ('person', )
BATCHSIZE_PER_GPU = 32
WORKERS = 8
dataset_type = 'CocoDataset'
train_path = "my_data_train.json"
val_path = "my_data_val.json"


# Run Time Setting 
max_epochs = 300
num_last_epochs = 15
resume_from = None

# For evaluation
eval_interval = 5
eval_metric = "bbox" # not 'bbox' | mAP
save_best_metric = "auto" # not 'auto' | mAP

# model settings
model = dict(
    type='mmdet.YOLOX',
    input_size=img_scale,
    random_size_range=(10, 20),
    random_size_interval=10,
    backbone=dict(
        type='MobileNetV2',
        widen_factor=0.5, 
        out_indices=(2, 4, 6),
        act_cfg=dict(type='ReLU6')),
    neck=dict(
        type='YOLOXPAFPN',
        in_channels=[16, 48, 160],
        out_channels=64,
        num_csp_blocks=1, 
        use_depthwise=True),
    bbox_head=dict(
        type='YOLOXHead', 
        num_classes=NC, 
        in_channels=64, 
        feat_channels=64, 
        use_depthwise=True),
    train_cfg=dict(assigner=dict(type='SimOTAAssigner', center_radius=2.5)),
    # In order to align the source code, the threshold of the val phase is
    # 0.01, and the threshold of the test phase is 0.001.
    test_cfg=dict(score_thr=0.01, nms=dict(type='nms', iou_threshold=0.65)))

# Training pipeline 
train_pipeline = [
    dict(type='Mosaic', img_scale=img_scale, pad_val=114.0),
    dict(
        type='RandomAffine',
        scaling_ratio_range=(0.5, 1.5),
        border=(-img_scale[0] // 2, -img_scale[1] // 2)),
    dict(type='YOLOXHSVRandomAug'),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(type='Resize', img_scale=img_scale, keep_ratio=True),
    dict(
        type='Pad',
        pad_to_square=True,
        pad_val=dict(img=(114.0, 114.0, 114.0))),
    dict(type='FilterAnnotations', min_gt_bbox_wh=(1, 1), keep_empty=False),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels'])
]

# Training dataset 
train_dataset = dict(
    type='MultiImageMixDataset',
    dataset=dict(
        type=dataset_type, 
        ann_file=train_path,
        classes=CLASSES, 
        img_prefix=None,
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(type='LoadAnnotations', with_bbox=True)
        ],
        filter_empty_gt=False,
    ),
    pipeline=train_pipeline)

# Test pipeline 
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=test_img_scale,
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(
                type='Pad',
                pad_to_square=True,
                pad_val=dict(img=(114.0, 114.0, 114.0))),
            dict(type='DefaultFormatBundle'),
            dict(type='Collect', keys=['img'])
        ])
]

data = dict(
    samples_per_gpu=BATCHSIZE_PER_GPU,
    workers_per_gpu=WORKERS,
    persistent_workers=True,
    train=train_dataset,
    val=dict(
        type=dataset_type, 
        ann_file=val_path,
        classes=CLASSES, 
        img_prefix=None,
        pipeline=test_pipeline),
    test=dict(
        type=dataset_type, 
        ann_file=val_path,
        classes=CLASSES, 
        img_prefix=None,
        pipeline=test_pipeline))


# Algorithm Setting 
algorithm = dict(
    type='AutoSlim',
    num_sample_training=4, 
    architecture=dict(type='MMDetArchitecture', model=model),
    pruner=dict(
        type='RatioPruner',
        except_start_keys=[
            "bbox_head.multi_level_conv_cls.0",
            "bbox_head.multi_level_conv_cls.1",
            "bbox_head.multi_level_conv_cls.2",
            "bbox_head.multi_level_conv_reg.0",
            "bbox_head.multi_level_conv_reg.1",
            "bbox_head.multi_level_conv_reg.2",
            "bbox_head.multi_level_conv_obj.0",
            "bbox_head.multi_level_conv_obj.1",
            "bbox_head.multi_level_conv_obj.2"
        ], 
        ratios=(2 / 12, 3 / 12, 4 / 12, 5 / 12, 6 / 12, 7 / 12, 8 / 12, 9 / 12, 10 / 12, 11 / 12, 1.0)),
    retraining=False,
    bn_training_mode=True,
    input_shape=None)


# default 8 gpu
optimizer = dict(
    type='SGD',
    lr=0.02,
    momentum=0.9,
    weight_decay=5e-4,
    nesterov=True,
    paramwise_cfg=dict(norm_decay_mult=0., bias_decay_mult=0.))
optimizer_config = None

# learning policy
lr_config = dict(
    _delete_=True,
    policy='YOLOX',
    warmup='exp',
    by_epoch=False,
    warmup_by_epoch=True,
    warmup_ratio=1,
    warmup_iters=5,  # 5 epoch
    num_last_epochs=num_last_epochs,
    min_lr_ratio=0.05)


runner = dict(type='EpochBasedRunner', max_epochs=max_epochs)

custom_hooks = [
    dict(
        type='YOLOXModeSwitchHook',
        num_last_epochs=num_last_epochs,
        priority=48),
    dict(
        type='SyncNormHook',
        num_last_epochs=num_last_epochs,
        interval=eval_interval,
        priority=48),
    dict(
        type='ExpMomentumEMAHook',
        resume_from=resume_from,
        momentum=0.0001,
        priority=49)
]
# How many epochs saving one model, -1 we never save the intermediate checkpoints 
checkpoint_config = dict(interval=-1)
evaluation = dict(
    save_best=save_best_metric,
    # The evaluation interval is 'interval' when running epoch is
    # less than ‘max_epochs - num_last_epochs’.
    # The evaluation interval is 1 when running epoch is greater than
    # or equal to ‘max_epochs - num_last_epochs’.
    interval=eval_interval,
    dynamic_intervals=[(max_epochs - num_last_epochs, 1)],
    metric=eval_metric)
log_config = dict(interval=50)

# NOTE: `auto_scale_lr` is for automatically scaling LR,
# USER SHOULD NOT CHANGE ITS VALUES.
# base_batch_size = (8 GPUs) x (8 samples per GPU)
auto_scale_lr = dict(base_batch_size=64)


use_ddp_wrapper = True

```

### My effort to solve the problem
* First, I try to use mmrazor trainign pipeline to train the max_model only, which means that I set self.pruner.set_max_channel() only in def train_step() method. And the model converges as normal. 
* Then I try to train only the max and min model in one iteration, and evaluate the max model only, I found that the model can not converge. 
* I found that in mmrazor, you did not apply out_mask for BN layers, which does not align with the original implementation of AutoSlim, and you only apply mask for Conv and Linear layers. My question is that: is it missing or not necessary to apply out_mask for BN layers? I think this will make the update of BN layers incorrect. Can it be the reason why the training of yoloxnano does not converge? 

### My question:
* Any suggestion on using AutoSlim on yoloxnano?
* Regarding the third opinion above, any insight from your perspective? Is it simple a BUG or you got any trick which i did not find? 

Best regards

**Comment by HIT-cwh** [comment]
Hi! Thank you for your issue.

For example, here is a toy model:
```
class ToyModel(nn.Module):
    def __init__(self):
        super(ToyModel, self).__init__()
        self.conv1 = nn.Conv2d(2, 4, 1)
        self.bn = nn.BatchNorm2d(4)
        self.conv2 = nn.Conv2d(4, 2, 1)
        self.in_mask1 = torch.tensor([1, 1]).reshape(1, 2, 1, 1).bool()
        self.in_mask2 = torch.tensor([1, 1, 0, 0]).reshape(1, 4, 1, 1).bool()

    def forward(self, x):
        # input x (1, 2, 2, 2)
        x = self.conv1(x)  # x after conv1 (1, 4, 2, 2)
        x = self.bn(x)     # x after bn1 (1, 4, 2, 2)
        x = x * self.in_mask2   # x after in_mask2 (1, 4, 2, 2), (1, 2, 2, 2) in fact
        return self.conv2(x)  # x after conv2 (1, 2, 2, 2)
```

If we prune the last two channels of `conv1` and `bn`, we actually only need to mask the last two channels of `conv2`'s input. The weight and bias of `bn` module are correct as their gradients are 0 due to the `in_mask2`. As the `running_mean` and `running_var` are incorrect, we set ``nn.BatchNorm`` to training mode when the model is set to eval mode during evaluation. And they will be sliced during deployment. So it may not negatively interfere with the training phase of the model.

**Comment by TianpengBu** [comment]
Hi! Thanks for the ToyModel and explaination. It's quite clear that the weight and bias will not be updated after multiplying a mask. 

As I consider the first question, the validation mAP is still 0 while training, no matter I use `bn_training_mode=True` or `bn_training_mode=False`. Any idea regarding this question? If you need more training details, please let me know.

**Comment by TianpengBu** [comment]
> Hi! Thanks for the ToyModel and explaination. It's quite clear that the weight and bias will not be updated after multiplying a mask.
> 
> As I consider the first question, the validation mAP is still 0 while training, no matter I use `bn_training_mode=True` or `bn_training_mode=False`. Any idea regarding this question? If you need more training details, please let me know.
